trajoptlib/src/mecanum_trajectory_generator.cpp

```diff
@@ -0,0 +1,449 @@
+    x.emplace_back(problem.decision_variable());
+    y.emplace_back(problem.decision_variable());
+    cosθ.emplace_back(problem.decision_variable());
+    sinθ.emplace_back(problem.decision_variable());
+    vx.emplace_back(problem.decision_variable());
+    vy.emplace_back(problem.decision_variable());
+    ω.emplace_back(problem.decision_variable());
+    ax.emplace_back(problem.decision_variable());
+    ay.emplace_back(problem.decision_variable());
+    α.emplace_back(problem.decision_variable());
+
+    for (size_t module_index = 0; module_index < module_cnt; ++module_index) {
+      F.at(index).emplace_back(problem.decision_variable());
+    }
+
+    dts.emplace_back(problem.decision_variable());
+  }
+
+  // Minimize total time
+  const double s_init = 1.0 / std::sqrt(2.0);
+  const double max_force =
+      path.drivetrain.wheel_max_torque * 4 / path.drivetrain.wheel_radius * s_init;
+  const auto max_accel = max_force / path.drivetrain.mass;
+  const double max_drivetrain_velocity =
+      path.drivetrain.wheel_radius * path.drivetrain.wheel_max_angular_velocity;
+  auto max_wheel_position_radius = 0.0;
+  for (auto module : path.drivetrain.wheels) {
+    max_wheel_position_radius =
+        std::max(max_wheel_position_radius, module.norm());
+  }
+  const auto max_ang_vel = max_drivetrain_velocity / max_wheel_position_radius;
+  const auto max_ang_accel = max_accel / max_wheel_position_radius;
+  for (size_t sgmt_index = 0; sgmt_index < Ns.size(); ++sgmt_index) {
+    auto N_sgmt = Ns.at(sgmt_index);
+    const auto sgmt_start = get_index(Ns, sgmt_index);
+    const auto sgmt_end = get_index(Ns, sgmt_index + 1);
+
+    if (N_sgmt == 0) {
+      for (size_t index = sgmt_start; index < sgmt_end + 1; ++index) {
+        dts.at(index).set_value(0.0);
+      }
+    } else {
+      // Use initial_guess and Ns to find the dx, dy, dθ between wpts
+      const auto dx =
+          initial_guess.x.at(sgmt_end) - initial_guess.x.at(sgmt_start);
+      const auto dy =
+          initial_guess.y.at(sgmt_end) - initial_guess.y.at(sgmt_start);
+      const auto dist = std::hypot(dx, dy);
+      const auto θ_0 = std::atan2(initial_guess.thetasin.at(sgmt_start),
+                                  initial_guess.thetacos.at(sgmt_start));
+      const auto θ_1 = std::atan2(initial_guess.thetasin.at(sgmt_end),
+                                  initial_guess.thetacos.at(sgmt_end));
+      const auto dθ = std::abs(angle_modulus(θ_1 - θ_0));
+
+      auto max_linear_vel = max_drivetrain_velocity;
+
+      const auto combined_time = std::max(
+          calculate_trapezoidal_time(dist, max_linear_vel, max_accel),
+          calculate_trapezoidal_time(dθ, max_ang_vel, max_ang_accel));
+      const double sgmt_time = combined_time;
+
+      for (size_t index = sgmt_start; index < sgmt_end + 1; ++index) {
+        auto& dt = dts.at(index);
+        problem.subject_to(dt >= 0);
+        problem.subject_to(dt <= 3);
```

Comment:
You'll need to merge main to gain access to this new function.
```suggestion
        problem.subject_to(slp::bounds(0, dt, 3));
```